.github/workflows/release.yml

```diff
@@ -63,26 +161,49 @@ jobs:
 
       - run: python -m pip install --upgrade pip wheel
       - run: python setup.py sdist
-
-      - run: python -m pip install dist/chromo-*
+      - run: python -m pip install --prefer-binary -v `echo dist/chromo-*`[test]
       - run: python -m pip install pytest
-      - run: python -m pytest
+      - run: python -m pytest -vv
 
       - uses: actions/upload-artifact@v3
         with:
           path: dist/*.tar.gz
 
-  upload:
+  upload_to_PyPI:
+    name: Upload to PyPI
     needs: [wheels, sdist]
     runs-on: ubuntu-latest
-    if: ${{ github.ref == 'refs/heads/master' }}
+    if: ${{ github.ref == 'refs/heads/main' && startsWith(github.ref, 'refs/tags/v') }}
     steps:
-      - uses: actions/download-artifact@v3
+      - name: Download files in artifact to "dist/" directory
+        uses: actions/download-artifact@v3
         with:
           name: artifact
           path: dist
 
-      - uses: pypa/gh-action-pypi-publish@release/v1
+      - name: Publish package to PyPI
+        uses: pypa/gh-action-pypi-publish@v1.6.4
         with:
           user: __token__
-          password: ${{secrets.pypi_password}}
+          password: ${{ secrets.PYPI_API_TOKEN }}
+
+
+  upload_to_TestPyPI:
+    name: Upload to TestPyPI
+    needs: [wheels, sdist]
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'workflow_dispatch' && inputs.upload_to_TestPyPI == true }}
+    steps:
+      - name: Download files in artifact to "dist/" directory
+        uses: actions/download-artifact@v3
+        with:
+          name: artifact
+          path: dist
+
+      - name: Publish package to TestPyPI
+        uses: pypa/gh-action-pypi-publish@v1.6.4
+        with:
+          user: __token__
+          repository_url: https://test.pypi.org/legacy/
+          password: ${{ secrets.TEST_PYPI_API_TOKEN }}
```

Comment:
You can remove this, it is not useful to have. Wheels are already automatically tested by some code further up, if wheels are broken, they are not uploaded to PyPI anyway. The only thing that can fail is the upload itself, but if that fails, nothing is released on PyPI and no harm is done. TestPyPI is only useful for people who do not release automatically via CI scripts.

Reply:
I see. I will leave it for a while until we try to upload to TestPyPI.  When we start to upload to PyPI I will remove it.

Reply:
It is better to test uploading directly on PyPI. The only thing that can fail at this point, since you know that the wheels work, is that the API token is wrong, and in that case you don't upload at all. The API tokens for TestPyPI and PyPI are different, so it is better to test directly on the real thing.

If you are uneasy to make the first release, then release a +dev version or a beta or a release candidate. This can be achieved by adding an appropriate suffix to the version number. PyPI will not resolve to these versions when someone types `pip install chromo`, so it is safe. You don't need TestPyPI.

Reply:
Ok, I will delete it.